File: test/unused/PFN_vkDebugReportCallbackEXT.java
package test.unused;

import com.sun.jna.Callback;
import com.sun.jna.Pointer;
import cz.mg.vulkan.jna.types.size_t;


/**
 *  typedef VkBool32 (VKAPI_PTR *PFN_vkDebugReportCallbackEXT)(
 *      VkDebugReportFlagsEXT                       flags,
 *      VkDebugReportObjectTypeEXT                  objectType,
 *      uint64_t                                    object,
 *      size_t                                      location,
 *      int32_t                                     messageCode,
 *      const char*                                 pLayerPrefix,
 *      const char*                                 pMessage,
 *      void*                                       pUserData
 *  );
 *  @see <a href="https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/PFN_vkDebugReportCallbackEXT.html">khronos documentation</a>
 **/
public interface PFN_vkDebugReportCallbackEXT extends Callback {
    public int PFN_vkDebugReportCallbackEXT(
        VkDebugReportFlagsEXT.ByValue flags,
        VkDebugReportObjectTypeEXT.ByValue objectType,
        long object,
        size_t location,
        int messageCode,
        String pLayerPrefix,
        String pMessage,
        Pointer pUserData
    );
}

File: c/cz/mg/vulkan/PFN_vkDebugReportCallbackEXT.c
#include <vulkan/vulkan.h>
#include <jni.h>
#include <stdlib.h>
#include <string.h>

jlong jniPointerToLong(const void* p);
void* jniLongToPointer(jlong l);
jlong jniFunctionPointerToLong(PFN_vkVoidFunction p);
PFN_vkVoidFunction jniLongToFunctionPointer(jlong l);
void jniThrowException(JNIEnv* env, const char* message);

void Java_cz_mg_vulkan_PFNvkDebugReportCallbackEXT_call(JNIEnv* env, jclass jc, jlong address, jlong flags, jlong objectType, jlong object, jlong location, jlong messageCode, jlong pLayerPrefix, jlong pMessage, jlong pUserData, jlong rval){
    (void)env;
    (void)jc;
    PFN_vkDebugReportCallbackEXT f = (PFN_vkDebugReportCallbackEXT)jniLongToFunctionPointer(address);
    VkBool32* rvalAddress = jniLongToPointer(rval);
    *rvalAddress = f(
        *((VkDebugReportFlagsEXT*)jniLongToPointer(flags)),
        *((VkDebugReportObjectTypeEXT*)jniLongToPointer(objectType)),
        *((uint64_t*)jniLongToPointer(object)),
        *((size_t*)jniLongToPointer(location)),
        *((int32_t*)jniLongToPointer(messageCode)),
        ((char*)jniLongToPointer(pLayerPrefix)),
        ((char*)jniLongToPointer(pMessage)),
        ((void*)jniLongToPointer(pUserData))
    );
}
